OSDOCS-6248:adds AdminNetworkPolicy to OVN-K (#70800)
openshift/openshift-docs: modules/nw-ovn-k-adminnetwork-policy.adoc

// Module included in the following assemblies:
//
// * networking/ovn-k-network-policy.adoc

:_mod-docs-content-type: <Concept>
[id="AdminNetworkPolicy"_{context}"]
= AdminNetworkPolicy

> `AdminNetworkPolicy` (ANP) is a cluster-scoped custom resource definitions (CRDs). As an {product-title} administrator, you can use `AdminNetworkPolicy` to secure your cluster in the following ways:

Reviewer: ```suggestion
`AdminNetworkPolicy` (ANP) is a cluster-scoped custom resource definition (CRD) that is maintained and shipped by sig-network-policy-api working group <link to https://github.com/kubernetes-sigs/network-policy-api>. As an {product-title} administrator, you can use `AdminNetworkPolicy` to secure your cluster's networking in the following ways:
```

Author: We have links to sig docs in other spaces so I'm just adding this to the Additional resources section.

Reviewer: ack I didn't find it anywhere, must have missed the link to the KEP. Could you please tell me where?

Author: https://docs.openshift.com/container-platform/4.14/machine_management/capi-machine-management.html

> * You can create cluster-scoped policies prior to namespaces being created.

Reviewer: This feels off to me, how about just say: You can create ANPs even before namespaces are created thus enforce security rules earlier on or something....
First point and second  point have no correlation IMHO..

* You can create deny and allow traffic rules at a cluster level that are non-overridable by users in the cluster.

The key difference between `AdminNetworkPolicy` and `NetworkPolicy` is that ANP provides cluster scoped policy for administrators while `NetworkPolicy` provides namespaced scoped policy for developers.

An ANP allows administrators to specify:

* A `subject` that consists of a set of pods that is selected by the policy

* A list of ingress rules to be applied for all ingress traffic towards the `subject`.

> * A list of egress rules to be applied for all egress traffic from the `subject`.

Reviewer: what about priorities in ANP? Are you covering that elsewhere? it also allows to specific a priority...

//this content may need to be moved to the assembly to help discuss ACLs with the image. This would provide a more detailed picture of the integration between ANP and BANP with NP API.

> If traffic matches an ANP rule, the rules in that ANP will be evaluated first. If the match is an ANP `allow` or `deny` rule, any existing `NetworkPolicies` and `BaselineAdminNetworkPolicy` (BANP) in the cluster will be intentionally skipped from evaluation. If the match is an ANP `pass` rule, then evaluation moves from tier 1 of the ACLs to tier 2 where `NetworkPolicy` is evaluated.

Reviewer: This has to go with tier-ing concept you mentioned above, this is completely off-track here... and feels disconnected....

[discrete]
[id="AdminNetworkPolicy-example"]
== AdminNetworkPolicy example

> `AdminNetworkPolicy` is a `TechnologyPreviewNoUpgrade` feature that can be enabled on test clusters that are not in production. The feature is behind the flag `enable-admin-network-policy`. See "Enabling features using feature gates" in the Additional resources of this section for more information on feature gates and `TechnologyPreviewNoUpgrade` features.

Reviewer: Why is this paragraph not a section in itself?
`Enabling ANP feature on OCP`

The title example is not suiting this paragraph, we must start with sample yamls straightway once we talk about examples in the heading right?

.Example YAML file for ANP
[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:

>   name: priority-50-example <1>

Reviewer: I would give this a better name that suits the purpose of the policy rather than repeating its priority number?

// Module included in the following assemblies:
//
// * networking/ovn-k-network-policy.adoc

:_mod-docs-content-type: Concept
[id="AdminNetworkPolicy"_{context}"]
= AdminNetworkPolicy

> `AdminNetworkPolicy` (ANP) is a cluster-scoped custom resource definition (CRD). As a {product-title} administrator, you can use ANP to secure your cluster's networking in the following ways:

Reviewer 2: Maybe:
```suggestion
An `AdminNetworkPolicy` (ANP) is a cluster-scoped custom resource definition (CRD). As a {product-title} administrator, you can use ANP to secure your cluster's networking in the following ways:
```

> * You can create cluster-scoped policies that are enforced before namespaces are created.

Author: I'll push an update that address this.

> The key difference between `AdminNetworkPolicy` and `NetworkPolicy` objects are that the former is for administrators and is cluster scoped while the latter is for tenant owners and is namespace scoped.

Author: I saw that they had something like that in the upstream [docs](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2091-admin-network-policy#key-differences-between-adminnetworkpolicies-and-networkpolicies). I'm happy to bring that in here if @tssurya and you both think it is necessary.

* You can create deny and allow traffic rules at a cluster level that are non-overridable by users in the cluster.

> The key difference between `AdminNetworkPolicy` and `NetworkPolicy` is that ANP provides cluster scoped policy for administrators while `NetworkPolicy` provides namespaced scoped policy for developers.

Reviewer 2: ```suggestion
The key difference between `AdminNetworkPolicy` and `NetworkPolicy` resources is that an ANP provides cluster scoped policy for administrators while a `NetworkPolicy` resource provides namespaced scoped policy for developers.
```

> The key difference between `AdminNetworkPolicy` and `NetworkPolicy` is that ANP provides cluster scoped policy for administrators while `NetworkPolicy` provides namespaced scoped policy for developers.

Author: I like this better than what we came up with. Thanks

> An ANP allows administrators to specify:

Reviewer 2: ```suggestion
An ANP allows administrators to specify the following:
```

* A `subject` that consists of a set of pods that is selected by the policy

> * A list of ingress rules to be applied for all ingress traffic towards the `subject`.
> 
> * A list of egress rules to be applied for all egress traffic from the `subject`.

Reviewer 2: ```suggestion
* A list of ingress rules to be applied for all ingress traffic towards the `subject`

* A list of egress rules to be applied for all egress traffic from the `subject`
```

> * A list of ingress rules to be applied for all ingress traffic towards the `subject`.
> 
> * A list of egress rules to be applied for all egress traffic from the `subject`.
> 
> * A `priority` value that determines the order of its evaluation.

Reviewer 2: ```suggestion
* A list of ingress rules to be applied for all ingress traffic towards the `subject`

* A list of egress rules to be applied for all egress traffic from the `subject`

* A `priority` value that determines the order of its evaluation
```

[NOTE]
====

> `AdminNetworkPolicy` is a `TechnologyPreviewNoUpgrade` feature that can be enabled on test clusters that are not in production. The feature is behind the flag `enable-admin-network-policy`. See "Enabling features using feature gates" in the Additional resources of this section for more information on feature gates and `TechnologyPreviewNoUpgrade` features.

Reviewer 2: ```suggestion
The `AdminNetworkPolicy` resource is a `TechnologyPreviewNoUpgrade` feature that can be enabled on test clusters that are not in production. The feature is behind the flag `enable-admin-network-policy`. For more information on feature gates and `TechnologyPreviewNoUpgrade` features, see "Enabling features using feature gates" in the "Additional resources" of this section.
```

> `AdminNetworkPolicy` is a `TechnologyPreviewNoUpgrade` feature that can be enabled on test clusters that are not in production. The feature is behind the flag `enable-admin-network-policy`. See "Enabling features using feature gates" in the Additional resources of this section for more information on feature gates and `TechnologyPreviewNoUpgrade` features.

Author: We cannot stack admonitions like that in our docs. `Keep admonitions to a minimum, and avoid placing multiple admonitions close to one another. If multiple admonitions are necessary, restructure the information by moving the less-important statements into the flow of the main content.` [Here](https://redhat-documentation.github.io/supplementary-style-guide/#admonitions:~:text=Keep%20admonitions%20to%20a%20minimum%2C%20and%20avoid%20placing%20multiple%20admonitions%20close%20to%20one%20another.%20If%20multiple%20admonitions%20are%20necessary%2C%20restructure%20the%20information%20by%20moving%20the%20less%2Dimportant%20statements%20into%20the%20flow%20of%20the%20main%20content.)

The [NOTE] seems less important than the [IMPORTANT] and we always begin TP features with that snippet for px support reasons.

> ====
> [discrete]

Reviewer 2: Nit, probably add a blank line here:
```suggestion
====

[discrete]
```

> [id="AdminNetworkPolicy-example"]

Reviewer 2: ```suggestion
[id="adminnetworkpolicy-example_{context}"]
```

== AdminNetworkPolicy example

> .Example YAML file for ANP

Reviewer 2: ```suggestion
.Example YAML file for an ANP
```

[%collapsible]
====
[source,yaml]
----
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:

>   name: Deny_ig_Pass_eg_example <1>

Reviewer: ```suggestion
  name: sample-anp-using-deny-pass-rules <1>
```

// Module included in the following assemblies:
//
// * networking/ovn-k-network-policy.adoc

:_mod-docs-content-type: CONCEPT
[id="adminnetworkpolicy_{context}"]
= AdminNetworkPolicy

An `AdminNetworkPolicy` (ANP) is a cluster-scoped custom resource definition (CRD). As a {product-title} administrator, you can use ANP to secure your network by creating network policies before creating namespaces. Additionally, you can create network policies on a cluster-scoped level that is non-overridable by `NetworkPolicy` objects.

The key difference between `AdminNetworkPolicy` and `NetworkPolicy` objects are that the former is for administrators and is cluster scoped while the latter is for tenant owners and is namespace scoped.

An ANP allows administrators to specify the following:

* A `priority` value that determines the order of its evaluation. The lower the value the higher the precedence.

> * A `subject` that consists of a set of pods that is selected by the policy.

Reviewer 3: Subject has namespace so set of pods that are in the selected namespace.
<pre>
subject:
    namespaces:
      matchLabels:
          kubernetes.io/metadata.name: test
</pre>

Reviewer 3: * A `subject` that consists of a set of namespaces or namespace. is what you have for BANP.

Author: Should it be namespaces, technically and no namespace (singular)?

Reviewer 3: namespaces: {} refers to all so it can be namespaces or namespace as narrowed down with matchLabels.
Also same label can be applied to more than one namespace. 
Therefore **that consists of a set of namespaces or namespace** is good.

* A list of ingress rules to be applied for all ingress traffic towards the `subject`.

* A list of egress rules to be applied for all egress traffic from the `subject`.

[NOTE]
====

> The `AdminNetworkPolicy` resource is a `TechnologyPreviewNoUpgrade` feature that can be enabled on test clusters that are not in production. The feature is behind the flag `enable-admin-network-policy`. For more information on feature gates and `TechnologyPreviewNoUpgrade` features, see "Enabling features using feature gates" in the "Additional resources" of this section.

Reviewer 3: @JoeAldinger From where does this flag **enable-admin-network-policy** come.
oc get featuregate cluster -oyaml
<pre>
.....
spec:
  featureSet: TechPreviewNoUpgrade
status:
  featureGates:
  - disabled:
    - name: ClusterAPIInstall
    - name: DisableKubeletCloudCredentialProviders
    - name: EventedPLEG
    - name: MachineAPIOperatorDisableMachineHealthCheckController
    enabled:
    - name: AdminNetworkPolicy
    - name: AlibabaPlatform
    - name: AutomatedEtcdBackup
    - name: AzureWorkloadIdentity
....
</pre>

Author: Isn't it listed right there. Or is that saying is isn't behind a feature gate and is default enabled.

Reviewer 3: name is AdminNetworkPolicy. Not sure I have seen enable-admin-network-policy. Also you enable all techpreview features not specifically one of them.

Author: Ok, I see what you are saying now. I can clean this up. Thank you.

  subject:
    namespaces:
      matchLabels:
          kubernetes.io/metadata.name: example.name <3>
  ingress: <4>
  - name: "deny-all-ingress-tenant-1" <5>
    action: "Deny"
    from:
    - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              custom-anp: tenant-1 <6>
        podSelector:
          matchLabels:
            custom-anp: tenant-1 <7>
  egress:<8>
  - name: "pass-all-egress-to-tenant-1"
    action: "Pass"
    to:
    - pods:
        namespaces:
          namespaceSelector:
            matchLabels:
              custom-anp: tenant-1
        podSelector:
          matchLabels:
            custom-anp: tenant-1
----
<1> Specify a name for your ANP.
<2> The `spec.priority` field supports a maximum of 100 ANP in the values of 0-99 in a cluster. The lower the value the higher the precedence. Creating `AdminNetworkPolicy` with the same priority creates a nondeterministic outcome.
<3> Specify the namespace to apply the ANP resource.
<4> ANP have both ingress and egress rules. ANP rules for `spec.ingress` field accepts values of `Pass`, `Deny`, and `Allow` for the `action` field.
<5> Specify a name for the `ingress.name`.

> <6> Specify the namespaces to select the pods from to apply the ANP resource.

Reviewer 3: Should the <6> be next to namespaces field instead of matchLabels.
Technically it could be matchLabels or matchExpressions for namespaceSelector.
